pkg/sql/colexec/execgen/inline.go

```diff
@@ -0,0 +1,363 @@
+	if decl.Type.Results == nil {
+		return &dst.EmptyStmt{}, nil
+	}
+	results := decl.Type.Results.List
+	retValNames = make([]string, len(results))
+	specs := make([]dst.Spec, len(results))
+	for i, result := range results {
+		var retvalName string
+		// Make a mangled name.
+		if len(result.Names) == 0 {
+			retvalName = fmt.Sprintf("__retval_%d", i)
+		} else {
+			retvalName = fmt.Sprintf("__retval_%s", result.Names[0])
+		}
+		retValNames[i] = retvalName
+		specs[i] = &dst.ValueSpec{
+			Names: []*dst.Ident{dst.NewIdent(retvalName)},
+			Type:  dst.Clone(result.Type).(dst.Expr),
+		}
+	}
+	retVal := &dst.DeclStmt{
+		Decl: &dst.GenDecl{
+			Tok:   token.VAR,
+			Specs: specs,
+		},
+	}
+	return retVal, retValNames
+}
+
+// getFormalParamReassignments creates a new DEFINE (:=) statement per parameter
+// to a FuncDecl, which makes a fresh variable with the same name as the formal
+// parameter name and assigns it to the corresponding name in the CallExpr.
+//
+// For example, given a FuncDecl:
+//
+// func foo(a int, b string) { ... }
+//
+// and a CallExpr
+//
+// foo(x, y)
+//
+// we'll return the statement:
+//
+// var (
+//   a int = x
+//   b string = y
+// )
+//
+// In the case where the formal parameter name is the same as the the input
+// parameter name, no extra assignment is created.
+func getFormalParamReassignments(decl *dst.FuncDecl, callExpr *dst.CallExpr) dst.Stmt {
+	formalParams := decl.Type.Params.List
+	reassignmentSpecs := make([]dst.Spec, 0, len(formalParams))
+	for i, formalParam := range formalParams {
+		if inputIdent, ok := callExpr.Args[i].(*dst.Ident); ok && inputIdent.Name == formalParam.Names[0].Name {
+			continue
+		}
+		reassignmentSpecs = append(reassignmentSpecs, &dst.ValueSpec{
+			Names:  []*dst.Ident{dst.NewIdent(formalParam.Names[0].Name)},
+			Type:   dst.Clone(formalParam.Type).(dst.Expr),
+			Values: []dst.Expr{callExpr.Args[i]},
+		})
+	}
+	if len(reassignmentSpecs) == 0 {
+		return &dst.EmptyStmt{}
+	}
+	return &dst.DeclStmt{
+		Decl: &dst.GenDecl{
+			Tok:   token.VAR,
+			Specs: reassignmentSpecs,
+		},
+	}
+}
+
+// replaceReturnStatements edits the input BlockStmt, from the function funcName,
+// replacing ReturnStmts at the end of the BlockStmts with the results of
+// applying returnEditor on the ReturnStmt or deleting them if the modifier is
+// nil.
+// It will panic if any return statements are not in the final position of the
+// input block.
+func replaceReturnStatements(
+	funcName string, stmt *dst.BlockStmt, returnModifier func(*dst.ReturnStmt) dst.Stmt,
+) *dst.BlockStmt {
+	// Remove return values if there are any, since we're ignoring returns
+	// as a raw function call.
+	var seenReturn bool
+	return dstutil.Apply(stmt, func(cursor *dstutil.Cursor) bool {
+		if seenReturn {
+			panic(fmt.Errorf("can't inline function %s: return not at end of body (found %s)", funcName, cursor.Node()))
+		}
+		n := cursor.Node()
+		switch t := n.(type) {
+		case *dst.FuncLit:
```

Comment:
Can you add what go code matches a `FuncLit`?

Reply:
Done, it's a function literal, like `x := func() bool {return true}`